### PR TITLE
make modulated_deform_conv_backward faster on camb

### PR DESCRIPTION
**Pull request by wyz5864**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu [hunk 1]
```diff
@@ -26,36 +26,42 @@ __mlu_func__ void dmcn_im2col_bilinear(const T *input, const int data_width,
     T* v2 = v1 + channel_per_group_align;
     T* v3 = v2 + channel_per_group_align;
     T* v4 = v3 + channel_per_group_align;
-    __nramset(v1, channel_per_group_align, (T)0);
-    __nramset(v2, channel_per_group_align, (T)0);
-    __nramset(v3, channel_per_group_align, (T)0);
-    __nramset(v4, channel_per_group_align, (T)0);
+
+    T w1 = hh * hw, w2 = hh * lw, w3 = lh * hw, w4 = lh * lw;
```
Comment (chenyuwang814) [inline mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:30]: 这个为啥要挪上来

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu [hunk 2]
```diff
@@ -133,15 +135,15 @@ __mlu_func__ void dmcn_get_coordinate_weight(T argmax_h, T argmax_w, const int h
         }
         __asm__ volatile("sync;");
         if (argmax_h_low >= 0 && argmax_w_low >= 0) {
-            __bang_mul_const(v1, v1, -1 * (argmax_w_low + 1 - argmax_w), channel_per_group_align);
```
Comment (chenyuwang814) [inline mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:136]: 这块原本-1 *是出于什么考虑有没有调研过

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu [hunk 3]
```diff
@@ -198,51 +200,52 @@ __mlu_func__ void modulated_deformable_im2col_mlu_kernel(
...
+            const T w_im = w_in + j * dilation_w + offset_w;
+
+            T* nram_val = (T*)nram_buffer;
+            T* nram_temp = nram_val + channel_per_group_align;
+            __nramset(nram_val, channel_per_group_align, (T)0);
+            if (h_im > -1 && w_im > -1 && h_im < height && w_im < width) {
+                dmcn_im2col_bilinear(data_im_ptr, width, height, width, h_im, w_im, c_im, num_channels,
+                                        nram_val, nram_temp, channel_per_deformable_group, channel_per_group_align);
             }
+            __bang_mul_const(nram_val, nram_val, mask, channel_per_group_align);
```
Comment (chenyuwang814) [inline mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:246]: __bang_mul_scalar

File: mmcv/ops/csrc/parrots/mlu/modulated_deform_conv_mlu.cpp [hunk 4]
```diff
@@ -399,11 +399,12 @@ void modulated_deform_conv_backward_camb_parrots(
     DArrayLite grad_input_nhwc = ctx.createDArrayLite(grad_input.spec().duplicate(parrots::MemoryFormat::ChannelsLast));
     cambTransposeTo(ctx, grad_input, grad_input_nhwc, CNNL_LAYOUT_NCHW, CNNL_LAYOUT_NHWC);
 
+    cnrtDim3_t k_dim = {getDeviceAttr(cnrtAttrMcorePerCluster), getDeviceAttr(cnrtAttrClusterCount), 1};
+    cnrtFunctionType_t k_type = CNRT_FUNC_TYPE_BLOCK;
```
Comment (chenyuwang814) [inline mmcv/ops/csrc/parrots/mlu/modulated_deform_conv_mlu.cpp:403]: 这里改成BLOCK的原因是啥

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu [hunk 5]
```diff
@@ -198,51 +200,52 @@ __mlu_func__ void modulated_deformable_im2col_mlu_kernel(
         const int width_col, T *data_col) {
     uint32_t idx = taskId;
     const int hw_col = height_col * width_col;
-    if (idx >= hw_col) return;
```
Comment (chenyuwang814) [inline mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:201]: 这里是原本return的逻辑错了吗，另外上面的core_nums参数是不是用不到了

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu [hunk 6]
```diff
@@ -198,51 +200,52 @@ __mlu_func__ void modulated_deformable_im2col_mlu_kernel(
...
-    for (int i = 0; i < kernel_h; ++i) {
-        for (int j = 0; j < kernel_w; ++j) {
-            for(int deformable_group_index = 0; deformable_group_index < deformable_group; deformable_group_index++) {
-                int c_im = deformable_group_index * channel_per_deformable_group;
-                const T *data_offset_ptr =
-                    data_offset + deformable_group_index * 2 * kernel_num * hw_col;
-                const T *data_mask_ptr =
-                    data_mask + deformable_group_index * kernel_num * hw_col;
-
-                for (uint32_t index = idx; index < hw_col; index += core_nums) {
```
Comment (chenyuwang814) [inline mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:217]: 这个里面是不是有大量的重复计算

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu (hunk 1 above)
Comment (wyz5864) [reply mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:30]: 之前有过一个改动方案需要，然后并不能提速，就没改回去

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu (hunk 2 above)
Comment (wyz5864) [reply mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:136]: 没有 我猜这个应该性能影响不大所以写的人没在意

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu (hunk 5 above)
Comment (wyz5864) [reply mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:201]: 是用不到了 改成内置的taskDim了 提这版代码的时候函数签名没改 我猜性能影响不大（也可能编译器会优化）

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu (hunk 6 above)
Comment (wyz5864) [reply mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:217]: 没太明白 具体指的是？

File: mmcv/ops/csrc/parrots/mlu/modulated_deform_conv_mlu.cpp (hunk 4 above)
Comment (wyz5864) [reply mmcv/ops/csrc/parrots/mlu/modulated_deform_conv_mlu.cpp:403]: 看文档的时候以为UNION1会只占用一个cluster 实测这样可以稍微快一点 就保留了

**Review by wzh1994 — APPROVED** [review]

**Review by mrdanielw — APPROVED** [review]

File: mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu (hunk 6 above)
Comment (chenyuwang814) [reply mmcv/ops/csrc/common/mlu/modulated_deform_conv_mlu_kernel.mlu:217]: 已讨论过，没有问题，是我一开始理解有误

**Pull request merged**